crates/btc-relay/src/lib.rs

```diff
@@ -151,131 +143,27 @@ pub mod pallet {
         ///
         /// * `block_header` - Bitcoin block header.
         ///
-        /// # <weight>
-        /// Key: C (len of chains), P (len of positions)
-        /// - Storage Reads:
-        /// 	- One storage read to check that parachain is not shutdown. O(1)
-        /// 	- One storage read to check if relayer authorization is disabled. O(1)
-        /// 	- One storage read to check if relayer is authorized. O(1)
-        /// 	- One storage read to check if block header is stored. O(1)
-        /// 	- One storage read to retrieve parent block hash. O(1)
-        /// 	- One storage read to check if difficulty check is disabled. O(1)
-        /// 	- One storage read to retrieve last re-target. O(1)
-        /// 	- One storage read to retrieve all Chains. O(C)
-        /// - Storage Writes:
-        ///     - One storage write to store block hash. O(1)
-        ///     - One storage write to store block header. O(1)
-        /// 	- One storage mutate to extend main chain. O(1)
-        ///     - One storage write to store best block hash. O(1)
-        ///     - One storage write to store best block height. O(1)
-        /// - Notable Computation:
-        /// 	- O(P) sort to reorg chains.
-        /// - Events:
-        /// 	- One event for block stored (fork or extension).
-        ///
-        /// Total Complexity: O(C + P)
-        /// # </weight>
+        /// ## Complexity
+        /// - `O(R)` where `R` is the number of reorgs
         #[pallet::call_index(1)]
-        #[pallet::weight(<T as Config>::WeightInfo::store_block_header())]
+        #[pallet::weight((
+            <T as Config>::WeightInfo::store_block_header(*reorg_bound),
+            DispatchClass::Operational
+        ))]
         #[transactional]
-        pub fn store_block_header(origin: OriginFor<T>, mut block_header: BlockHeader) -> DispatchResultWithPostInfo {
+        pub fn store_block_header(
+            origin: OriginFor<T>,
+            mut block_header: BlockHeader,
+            reorg_bound: u32,
```

Comment:
How do the clients determine what value to pass here?

Reply:
I added a note however that since `Chains` is unbounded this may become prohibitively expensive. We can resolve this by deleting old forks (also requires migrating `Chains` to a sorted vec instead of a map) but we should also consider weight refunding on more optimal complexity.

Reply:
>  but we should also consider weight refunding on more optimal complexity.
We already fully refund the fees for successful calls. The problem is that the majority of the calls fail, and you can't refund those (unless we start returning Ok for duplicate submissions, but that create a whole new problem)